### PR TITLE
Enable accessing fields that are classes without copying.

### PR DESCRIPTION
**Pull request by brendandahl**
Previously, fields that were classes were copied each time they were
accessed from JS. This led to some unexpected behavior (see #6402 and
#17573).

This is a revival of PR #6402. However, I opted to not add a new method
and instead add a new template specialization for `property` for fields
that are classes.

Fixes #6402, #17573

### CODE REVIEW

File: test/test_core.py [hunk 1]
```diff
@@ -7491,6 +7491,10 @@ def test_embind_negative_constants(self):
     self.emcc_args += ['-lembind']
     self.do_run_in_out_file_test('embind/test_negative_constants.cpp')
 
+  def test_embind_field_access(self):
+    self.emcc_args += ['-lembind']
+    self.do_run_in_out_file_test('embind/test_field_access.cpp', assert_identical=True)
```
Comment (sbc100) [inline test/test_core.py:7496]: Most tests don't use `assert_identical=True` .. is there some reason to use it here?

File: site/source/docs/porting/connecting_cpp_and_javascript/embind.rst [hunk 2]
```diff
@@ -199,6 +199,63 @@ For other code, that is not optimized by closure compiler, you do not need to
...
+
+    class SubClass {
+    public:
+      int i;
+    };
+
+    class MyClass {
+    public:
+      int x;
+      SubClass subClass;
```
Comment (sbc100) [inline site/source/docs/porting/connecting_cpp_and_javascript/embind.rst:218]: subclass is kind of a misnomer here, no?   InnerClass isn't much better.. NestedClass?   How about just `Foo`?

**Review by sbc100 — COMMENTED** [review]
My template-foo seems insufficient to review this.. net effect seems correct though!

**Comment by brendandahl** [comment]
I missed the specialization for `std::string` that was causing failures. I've added a check for that now ( `is_basic_string` ), that will prevent this new code path from happening for strings.

**Review by sebastiendu — APPROVED** [review]
Works for me !

**Comment by brendandahl** [comment]
I'm currently holding off on merging this. I'm hoping to come up with a more uniform plan for how embind handles references. I think ideally it'd be up to the user to decide if they should be by copy or directly modifiable.

**Comment by brendandahl** [comment]
Closing for now until we have a object ownership policy.

**Pull request closed without merging**